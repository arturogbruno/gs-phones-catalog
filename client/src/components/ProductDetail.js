import React from "react";
import { Link } from "react-router-dom";
import { useSelector } from "react-redux";
import CircleLoader from "react-spinners/CircleLoader";
import NotFound from "./NotFound";

const ProductDetail = ({ match }) => {
    const id = +match.params.id;
    const products = useSelector(state => state.products);
    
    if(products.length === 0) {
        return (
            <CircleLoader 
                color={"#00b091"}
                size={50}
            />
        )
    }
    
    const selectedProduct = products.find(product => product.id === id);

    if(selectedProduct) {
        const { manufacturer, name, price, imageFileName, description, processor, ram, screen, color } = selectedProduct;

        return (
            <div className="ProductDetail"> 
                <h1>This is the product detail page</h1>
                <Link to="/"><span role="img" aria-label="back link">⏪</span> Back</Link>
                <img src={`../images/${imageFileName}`} alt={`${manufacturer} ${name}`} />
                <h3>{manufacturer} {name}</h3>
                <h3>{price} €</h3>
                <p>{description}</p>
                <h5>Main features:</h5>
                <ul>
                    <li>Processor: {processor}</li>
                    <li>RAM: {ram} GB</li>
                    <li>Screen: {screen}</li>
                    <li>Color: {color}</li>
                </ul>
            </div>
        )
    } else {
        return <NotFound />
    }
}

export default ProductDetail;