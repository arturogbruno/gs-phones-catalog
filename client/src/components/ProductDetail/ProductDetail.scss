@import "../../styles/mixins.scss";

.ProductDetail {
    background: linear-gradient(180deg, rgba(255,255,255,1) 70%, rgba(255,255,255,0.5) 90%, rgba(255,255,255,0.5) 100%);
    color: rgb(35, 35, 35); 
    width: 80%;
    display: flex;
    flex-direction: column;
    margin-bottom: 100px;
    
    .productData {
        display: flex;
        flex-direction: column;

        .productImage {
            width: 100%;
            display: flex;
            justify-content: center;

            img {
                max-width: 100%;
                min-width: 100%;

                @include lg {
                    object-fit: contain;
                }
            }

            @include lg {
                width: 50%;
            }
        }

        .productInfo {
            padding: 5%;
            
            .productMainInfo {
                text-align: center;
                display: flex;
                flex-direction: column;
                margin-bottom: 50px;

                .price {
                    margin-top: 30px;
                    font-style: italic;
                }
            }

            .productSpecs {
                line-height: 2em;

                li {
                    list-style-type: square;
                    margin-left: 30px;
                }
            }

            @include lg {
                width: 50%;
            }
        }

        @include lg {
            height: 70%;
            flex-direction: row;
        }
    }

    .productDescription {
        height: 30%;
        padding: 0 5% 5% 5%;

        p {
            margin: 15px 0 30px 0;
            line-height: 1.3em;
        }

        button {
            border: 1px solid rgb(3, 185, 153);
            outline: none;
            background-color: transparent;
            padding: 5px 10px;
            transition-duration: 0.4s;
              
            &:hover {
                background-color: rgb(3, 185, 153);
                cursor: pointer;
            }
            
            a {
                text-decoration: none;
                font-size: 1.2em;
                color: rgb(35, 35, 35);
    
            }    
        }
    }

    @include lg {
        width: 60%;
        height: 70vh;
    }
}