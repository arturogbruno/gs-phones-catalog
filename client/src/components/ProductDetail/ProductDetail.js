import React from "react";
import { Link } from "react-router-dom";
import { useSelector } from "react-redux";
import CircleLoader from "react-spinners/CircleLoader";
import NotFound from "../NotFound/NotFound";
import "./ProductDetail.scss";

const ProductDetail = ({ match }) => {
    const id = +match.params.id;
    const products = useSelector(state => state.products);
    
    if(products.length === 0) {
        return (
            <CircleLoader 
                color={"#fff"}
                size={50}
            />
        )
    }
    
    const searchProduct = products => products.find(product => product.id === id);
    const foundProduct = searchProduct(products);

    if(foundProduct) {
        const { manufacturer, name, price, imageFileName, description, processor, ram, screen, color } = foundProduct;

        return (
            <div className="ProductDetail">
                <section className="productData">
                    <div className="productImage">
                        <img src={`../images/${imageFileName}`} alt={`${manufacturer} ${name}`} />
                    </div>
                    <div className="productInfo">
                        <div className="productMainInfo">
                            <h2>{manufacturer} {name}</h2>
                            <h2 className="price">{price} â‚¬</h2>
                        </div>
                        <div className="productSpecs">
                            <h4>Main features:</h4>
                            <ul>
                                <li>Processor: {processor}</li>
                                <li>RAM: {ram} GB</li>
                                <li>Screen: {screen}</li>
                                <li>Color: {color}</li>
                            </ul>
                        </div>
                    </div>
                </section>
                <section className="productDescription">
                    <h4>Description:</h4>
                    <p>{description}</p>
                    <Link to="/">Back</Link>
                </section>
            </div>
        )
    } else {
        return <NotFound />
    }
}

export default ProductDetail;