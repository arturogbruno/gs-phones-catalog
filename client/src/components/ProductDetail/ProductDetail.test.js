import React from "react";
import { BrowserRouter } from "react-router-dom";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import ProductDetail from "./ProductDetail";

describe("testing ProductDetail component", () => {
    let wrapper;
    let store;
    
    beforeEach(() => {
        const match = {
            params: {
                id: "2"
            }
        };

        const mockStore = configureMockStore();

        store = mockStore({
            products: [
                { 
                    id: 2,
                    name: "iPhone 7"
                }
            ]
        });

        wrapper = mount(
            <Provider store={store}>
                <BrowserRouter>
                    <ProductDetail match={match} />
                </BrowserRouter>
            </Provider>
        );
    });

    it("should render ProductDetail component", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("should render ProductDetail and productData elements if product is found", () => {
        expect(wrapper.find("ProductDetail").length).toEqual(1);
        expect(wrapper.find(".productData").exists()).toBe(true);
    });

    it("should render NotFound component if product is not found", () => {
        const match = {
            params: {
                id: "13"
            }
        };
        
        wrapper = mount(
            <Provider store={store}>
                <BrowserRouter>
                    <ProductDetail match={match} />
                </BrowserRouter>
            </Provider>
        );

        expect(wrapper.find("NotFound").length).toEqual(1);
    });
});