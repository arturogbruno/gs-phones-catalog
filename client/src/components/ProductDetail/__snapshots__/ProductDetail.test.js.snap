// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ProductDetail component should render ProductDetail component 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <ProductDetail
        match={
          Object {
            "params": Object {
              "id": "2",
            },
          }
        }
      >
        <div
          className="ProductDetail"
        >
          <section
            className="productData"
          >
            <div
              className="productImage"
            >
              <img
                alt="undefined iPhone 7"
                src="../images/undefined"
              />
            </div>
            <div
              className="productInfo"
            >
              <div
                className="productMainInfo"
              >
                <h2>
                   
                  iPhone 7
                </h2>
                <h2
                  className="price"
                >
                   â‚¬
                </h2>
              </div>
              <div
                className="productSpecs"
              >
                <h4>
                  Main features:
                </h4>
                <ul>
                  <li>
                    Processor: 
                  </li>
                  <li>
                    RAM: 
                     GB
                  </li>
                  <li>
                    Screen: 
                  </li>
                  <li>
                    Color: 
                  </li>
                </ul>
              </div>
            </div>
          </section>
          <section
            className="productDescription"
          >
            <h4>
              Description:
            </h4>
            <p />
            <Link
              to="/"
            >
              <LinkAnchor
                href="/"
                navigate={[Function]}
              >
                <a
                  href="/"
                  onClick={[Function]}
                >
                  Back
                </a>
              </LinkAnchor>
            </Link>
          </section>
        </div>
      </ProductDetail>
    </Router>
  </BrowserRouter>
</Provider>
`;
